### PR TITLE
Rewrite emake makefile

### PR DESCRIPTION
**Pull request by JoshDreamland**

### CODE REVIEW

File: CommandLine/emake/Makefile [hunk 1]
```diff
@@ -1,35 +1,46 @@
...
-	
-.eobjs:
-	mkdir -p .eobjs/
 
+EXTRA_SOURCES := ../../CompilerSource/settings-parse/eyaml.cpp ../../CompilerSource/compiler/event_reader/event_parser.cpp
+EXTRA_SRC_DIR := ../../CompilerSource
+
+CXXFLAGS  := -I$(SRC_DIR) -I$(EXTRA_SRC_DIR) -Wall -Wextra -Wpedantic -g
+LDFLAGS   := $(OS_LIBS)
+TEST_LIBS := -lgtest_main
```
Comment (fundies) [inline CommandLine/emake/Makefile]: Shouldn't we be linking gtest to enigma instead? I thought there would be a tests mode similar to debug / compile

File: CommandLine/emake/Makefile [hunk 2]
```diff
@@ -1,35 +1,46 @@
-CXX=g++
-CXXFLAGS=-g -std=c++11 -c -Wall -I. -I../../CompilerSource
+BINARY := ../../emake
+SRC_DIR := .
+OBJ_DIR := .eobjs
 
 ifeq ($(OS),Windows_NT)
```
Comment (fundies) [inline CommandLine/emake/Makefile]: This if doesn't actually work I have a fix for it in my mac fix pr. I never set $OS anywhere and OS isn't actually Windows_NT in mingw's shell anymore

File: CommandLine/emake/Makefile [hunk 3]
```diff
@@ -1,35 +1,46 @@
...
-	LDFLAGS=-lboost_system-mt -Wl,--no-as-needed -Wl,-rpath=./ -lboost_program_options-mt -lboost_iostreams-mt -lboost_filesystem-mt -lboost_system-mt -lpthread
+	OS_LIBS=-lboost_system-mt -Wl,--no-as-needed -Wl,-rpath=./ -lboost_program_options-mt -lboost_iostreams-mt -lboost_filesystem-mt -lboost_system-mt -lpthread
 else
-	LDFLAGS=-lboost_system -Wl,--no-as-needed -Wl,-rpath=./ -lboost_program_options -lboost_iostreams -lboost_filesystem -lboost_system -lpthread -ldl
+	OS_LIBS=-lboost_system -Wl,--no-as-needed -Wl,-rpath=./ -lboost_program_options -lboost_iostreams -lboost_filesystem -lboost_system -lpthread -ldl
 endif
-	
-SOURCES=$(shell find . -name "*.cpp")
-OBJECTS=$(addprefix .eobjs/,$(SOURCES:.cpp=.o))
-EXECUTABLE=../../emake
```
Comment (fundies) [inline CommandLine/emake/Makefile]: perhaphs we should replace ../../ with and ENIGMA_ROOT var incase anyone ever tries to move this

**Comment by fundies** [comment]
Now, this is what *I* would call completly broken :P

File: CommandLine/emake/Makefile [hunk 4]
```diff
@@ -1,35 +1,46 @@
...
-eventsres:
-	$(CXX) $(CXXFLAGS) ../../CompilerSource/compiler/event_reader/event_parser.cpp -o .eobjs/event_parser.o
-	
-.eobjs:
-	mkdir -p .eobjs/
 
+EXTRA_SOURCES := ../../CompilerSource/settings-parse/eyaml.cpp ../../CompilerSource/compiler/event_reader/event_parser.cpp
+EXTRA_SRC_DIR := ../../CompilerSource
+
+CXXFLAGS  := -I$(SRC_DIR) -I$(EXTRA_SRC_DIR) -Wall -Wextra -Wpedantic -g
```
Comment (fundies) [inline CommandLine/emake/Makefile]: -std=c++11

**Comment by RobertBColton** [comment]
Since this has merged, the total length of a build has increased by about two minutes:
https://ci.appveyor.com/project/enigma-dev/enigma-dev/build/1.0.313
https://ci.appveyor.com/project/enigma-dev/enigma-dev/build/1.0.312

The log of the first job that builds emake has also increased by about 900 lines:
https://ci.appveyor.com/project/enigma-dev/enigma-dev/build/1.0.313/job/u5smv85owlhvvnxs#L1939
https://ci.appveyor.com/project/enigma-dev/enigma-dev/build/1.0.312/job/6v32gcxy364md53l#L1103

**Pull request merged**